src/ray/raylet/actor_registration.cc

```diff
@@ -40,13 +40,39 @@ const std::unordered_map<ActorHandleID, ActorRegistration::FrontierLeaf>
   return frontier_;
 }
 
-void ActorRegistration::ExtendFrontier(const ActorHandleID &handle_id,
-                                       const ObjectID &execution_dependency) {
+ObjectID ActorRegistration::ExtendFrontier(const ActorHandleID &handle_id,
+                                           const ObjectID &execution_dependency) {
   auto &frontier_entry = frontier_[handle_id];
+  // Release the reference to the previous cursor for this
+  // actor handle, if there was one.
+  ObjectID object_to_release;
+  if (!frontier_entry.execution_dependency.is_nil()) {
+    auto it = dummy_objects_.find(frontier_entry.execution_dependency);
+    RAY_CHECK(it != dummy_objects_.end());
+    it->second--;
+    RAY_CHECK(it->second >= 0);
+    if (it->second == 0) {
+      object_to_release = frontier_entry.execution_dependency;
+      dummy_objects_.erase(it);
+    }
+  }
+
   frontier_entry.task_counter++;
   frontier_entry.execution_dependency = execution_dependency;
   execution_dependency_ = execution_dependency;
-  dummy_objects_.push_back(execution_dependency);
+  // Add the reference to the new cursor for this actor handle.
+  dummy_objects_[execution_dependency]++;
+  return object_to_release;
+}
+
+void ActorRegistration::AddHandle(const ActorHandleID &handle_id,
+                                  const ObjectID &execution_dependency) {
+  if (frontier_.find(handle_id) == frontier_.end()) {
+    frontier_[handle_id] = FrontierLeaf{
+        .task_counter = 0, .execution_dependency = execution_dependency,
```

Comment:
Hmm, I wasn't aware of any pros/cons. Do you have any suggestions in mind?

Reply:
Ok, when we converted from C to C++, e.g., in #321, we had to make some changes like

https://github.com/ray-project/ray/pull/321/files#diff-450748b4523710897a16506dabc79950L573

and

https://github.com/ray-project/ray/pull/321/files#diff-303e99e77aad9f39624c3769dd615ea0L220

which switched away from using this syntax. I thought it wasn't valid C++, though if it's compiling in your PR maybe I'm remembering it incorrectly. I'm sure @mehrdadn knows the answer.

As for suggestions, I'd defer to other people here. If it's valid C++11 then it's probably fine especially since it's super simple. I guess the alternative would be to manually set the fields after the declaration or to define a simple constructor.

Reply:
These are called designated initializers in C. They were extensions I believe until C99? I had no idea but apparently they've finally made it into C++20. I'd avoid using them given they're not valid C++11.

Reply:
Thanks @mehrdadn!

Ok, in that case let's either do

```cpp
frontier_[handle_id] = FrontierLeaf();
frontier_[handle_id].task_counter = 0;
frontier_[handle_id].execution_dependency = execution_dependency;
```

or define a constructor.

Reply:
If you want to initialize fields externally I'd get a reference and avoid looking up the same object even time. If you make a constructor that should be fine too, though semantically it'd be creating a new object to move onto the old object rather than modifying it or constructing it in place. (Note that using `operator[]` to index into a map automatically calls the default constructor, so if you don't want to call that, you'd want to use `insert`.)

Reply:
Okay, thanks!